addons/purchase/migrations/9.0.1.2/post-migration.py

```diff
@@ -92,9 +96,188 @@ def account_properties(cr):
+def set_po_line_amounts(env):
+    """We replicate here the code of the compute function and set the values
+    finally via SQL for avoiding the trigger of the rest of the computed
+    fields that depends on these fields.
+
+    Replaces code from https://bit.ly/2rgpTTV
+    """
+    lines = env['purchase.order.line'].search([])
+    for line in lines:
+        taxes = line.taxes_id.compute_all(
+            line.price_unit, line.order_id.currency_id, line.product_qty,
+            product=line.product_id, partner=line.order_id.partner_id)
+        env.cr.execute("""
+            UPDATE purchase_order_line
+            SET price_tax = %s,
+                price_total = %s,
+                price_subtotal = %s
+            WHERE id = %s""", (
+                taxes['total_included'] - taxes['total_excluded'],
+                taxes['total_included'],
+                taxes['total_excluded'],
+                line.id,
+            )
+        )
+
+
+@openupgrade.logging()
+def set_po_line_computed_rest(env):
+    """Emulate the computation of the rest of the computed fields through
+    equivalent SQL queries.
+    """
+    uom_precision = env['decimal.precision'].precision_get('Product Unit of Measure')
+    # Fail if there are UoM mismatches
+    # Replace https://bit.ly/2s1ePtG
+    openupgrade.logged_query(
+        env.cr,
+        """
+            SELECT
+                pol.id AS order_line_id,
+                pol_u.id AS order_line_uom_id,
+                pol_u.category_id AS order_line_uom_category_id,
+                pol.product_id AS order_line_product_id,
+                ail.id AS invoice_line_id,
+                ail_u.id AS invoice_line_uom_id,
+                ail_u.category_id AS invoice_line_uom_category_id,
+                ail_pp.id AS invoice_line_product_id,
+                ail_pt.name AS invoice_line_product_template_name
+            FROM purchase_order_line pol
+                JOIN account_invoice_line ail ON
+                    pol.id = ail.purchase_line_id
+                JOIN product_product ail_pp ON ail_pp.id = ail.product_id
+                JOIN product_template ail_pt ON ail_pt.id = ail_pp.product_tmpl_id
+                JOIN account_invoice ai ON ai.id = ail.invoice_id
+                JOIN product_uom ail_u on (ail_u.id = ail.uom_id)
+                JOIN product_uom pol_u on (pol_u.id = pol.product_uom)
+            WHERE
+                ail_u.category_id != pol_u.category_id
+        """
+    )
+    uom_category_mismatches = env.cr.dictfetchall()
+    if uom_category_mismatches:
+        raise Exception(
+            "Found these mismatching UoM in related purchase.order.line "
+            "and account.invoice.line records: %s" %
+            pformat(uom_category_mismatches))
+    # Replace https://bit.ly/363UnqR
+    openupgrade.logged_query(
+        env.cr,
+        """
+            UPDATE purchase_order_line pol
+            SET qty_invoiced = sub.qty
+            FROM (
+                SELECT
```

Comment:
I know this should already be a huge improvement over the ORM, but in terms of SQL subqueries are relatively expensive. Did you consider using a temporary table instead of the subquery? See https://github.com/VanMoof/OpenUpgrade/commit/2d26cdf118a226855193cfdce94c0131841c4925#diff-d4a6069f9ea8152872ee07aa6bd111fbR480 for an example.

Reply:
The subquery in this case is needed to be populated only once, so I don't think we are going to get a lot of improvement. We will execute the script tomorrow and tell you back what are the obtained times and the number of records.

Reply:
Sure! This is non blocking on my part.

Reply:
Subqueries are expensive when in `SELECT` or `WHERE`, because they are (or could be, depending on postgres optimizations and configuration) executed row by row.

When they're in the `FROM` part, they're not so expensive.

Reply:
On our real DB, this is the result:
```
109182 rows affected after 0:00:07.030084
```
What means 7 seconds for more than 100k records, which legitimates that this subquery usage is sane and no need to bother with temporary tables.

Reply:
Absolutely! Thanks for the feedback.